src/report.js

```diff
@@ -0,0 +1,180 @@
+      agent: globals.httpsAgent,
+      timeout: config.networkTimeout
+    }, (res) => {
+      var apiResponse = ''
+
+      res.on('data', function (chunk) {
+        apiResponse += chunk
+      })
+
+      res.on('end', function () {
+        resolve({ status: res.statusCode, apiResponse })
+      })
+    }).on('error', (err) => {
+      reject(err)
+    })
+
+    req.write(JSON.stringify(requestBody))
+    req.end()
+  })
+}
+
+class Report {
+  constructor(config, context, startTime, metrics) {
+    this.initalPromises = {
+      statPromise: system.readstat('self'),
+      bootIdPromise: system.readbootid()
+    }
+
+    // flag on report sending status, reports are sent once
+    this.sent = false
+
+    this.config = config
+    this.context = context
+    this.startTime = startTime
+
+    // Populate initial report skeleton on construction
+    this.report = {
+      client_id: config.clientId,
+      installMethod: config.installMethod,
+      duration: undefined,
+      processId: globals.PROCESS_ID,
+      aws: {
+        functionName: context.functionName,
+        functionVersion: context.functionVersion,
+        awsRequestId: context.awsRequestId,
+        invokedFunctionArn: context.invokedFunctionArn,
+        logGroupName: context.logGroupName,
+        logStreamName: context.logStreamName,
+        memoryLimitInMB: context.memoryLimitInMB,
+        getRemainingTimeInMillis: undefined,
+        traceId: process.env._X_AMZN_TRACE_ID
+      },
+      environment: {
+        agent: {
+          runtime: 'nodejs',
+          version: globals.VERSION,
+          load_time: globals.MODULE_LOAD_TIME
+        },
+        nodejs: {
+          version: process.version,
+          memoryUsage: undefined
+        },
+        host: {
+          container_id: undefined
+        },
+        os: {},
+      },
+      errors: {},
+      coldstart: globals.COLDSTART,
+      custom_metrics: metrics,
+    }
+  }
+
+  send(err, callback) {
+
+    // Send report only once
+    if (this.sent) {
+      return
+    }
+    const self = this
+    const config = this.config
+    const context = this.context
+
+    // Add error to report if necessary
+    if (err) {
+      this.report.errors = ((err) => {
+        return {
+          name: err.name,
+          message: err.message,
+          stack: err.stack,
+          lineNumber: err.lineNumber,
+          columnNumber: err.columnNumber,
+          fileName: err.fileName
+        }
+      })((typeof(err) === 'string') ? new Error(err) : err)
+    }
+
+    // Resolve system promises/report data
+    Promise.all([
+      this.initalPromises.statPromise,
+      system.readstat('self'),
+      system.readstatus('self'),
+      this.initalPromises.bootIdPromise
+    ]).then((results) => {
+      const pre_proc_self_stat = results[0],
+        proc_self_stat = results[1],
+        proc_self_status = results[2],
+        boot_id = results[3]
+
+      const osStats = {
+          hostname: os.hostname(),
```

Comment:
probably worth making this a follow-up patch, but we could try promisifying fetching these values too. Dunno if it'd provide a benefit...